crates/matrix-sdk-base/src/read_receipts.rs

```diff
@@ -0,0 +1,396 @@
+use eyeball_im::Vector;
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use ruma::{
+    events::{
+        poll::{start::PollStartEventContent, unstable_start::UnstablePollStartEventContent},
+        receipt::{ReceiptThread, ReceiptType},
+        room::message::Relation,
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent, OriginalSyncMessageLikeEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, OwnedEventId, RoomId, UserId,
+};
+use tracing::{instrument, trace};
+
+use super::BaseClient;
+use crate::{error::Result, store::StateChanges, RoomInfo};
+
+/// Provider for timeline events prior to the current sync.
+pub trait PreviousEventsProvider: Send + Sync {
+    /// Returns the list of known timeline events, in sync order, for the given
+    /// room.
+    fn for_room(&self, room_id: &RoomId) -> Vector<SyncTimelineEvent>;
+}
+
+impl PreviousEventsProvider for () {
+    fn for_room(&self, _: &RoomId) -> Vector<SyncTimelineEvent> {
+        Vector::new()
+    }
+}
+
+#[instrument(skip_all, fields(room_id = %room_info.room_id))]
+pub(crate) async fn compute_notifications(
+    client: &BaseClient,
+    changes: &StateChanges,
+    previous_events_provider: &dyn PreviousEventsProvider,
+    new_events: &[SyncTimelineEvent],
+    room_info: &mut RoomInfo,
+) -> Result<()> {
+    let user_id = &client.session_meta().unwrap().user_id;
+    let prev_latest_receipt_event_id = room_info.read_receipts.latest_read_receipt_event_id.clone();
+
+    if let Some(receipt_event) = changes.receipts.get(room_info.room_id()) {
+        trace!("Got a new receipt event!");
+
+        // Find a private or public read receipt for the current user.
+        let mut receipt_event_id = None;
+        if let Some((event_id, receipt)) =
+            receipt_event.user_receipt(user_id, ReceiptType::ReadPrivate)
```

Comment:
This code always picks the private receipt if it is present, but the [spec says](https://spec.matrix.org/v1.5/client-server-api/#private-read-receipts):

> Between m.read and m.read.private, the receipt which is more “ahead” or “recent” is used when determining the highest read-up-to mark

Reply:
Pondered about this a bit, and what we get from the read receipts extension of sliding sync is a mapping of room id -> event id -> user id -> receipt. So to know which receipt is "ahead" in terms of sync order, we'd need to iterate over the events (with the same caveat that they could be from this sync, or a previous one, or a future one).

Would using the time of the receipt be good enough to compare them? (it's the read event timestamp, if I'm not mistaken, so that's forgeable since it's user-defined, but I don't see any big threat here)

Reply:
As I tried to describe [here](https://docs.google.com/presentation/d/1H1gxRmRFAm8d71hCILWmpOYezsvdlb7cB6ANl-20Gns/edit?usp=sharing) it's tricky to decide the order of messages, and it does matter. If you disagree with the server over which message is first, the server will discard the receipts you send (if it think it already has a later one) and you will end up in an inconsistent state between server and client, potentially causing stuck unreads.

The `ts` in the receipt is the time at which it was created (I am fairly sure?), so it doesn't help us know which events are read. Even if it's the `ts` of the read event, it still might disagree with the server's idea of what is first.

Reply:
> If you disagree with the server over which message is first, the server will discard the receipts you send (if it think it already has a later one) and you will end up in an inconsistent state between server and client, potentially causing stuck unreads.

Thanks, that's new information :smiling_face_with_tear: